baselines/pFedHN/pFedHN/client.py

```diff
@@ -0,0 +1,178 @@
+"""Client Handling."""
+
+from collections import OrderedDict
+
+import flwr as fl
+import torch
+
+from pFedHN.models import CNNTarget
+from pFedHN.trainer import train
+from pFedHN.utils import get_device
+
+
+# pylint: disable=too-many-instance-attributes
+class FlowerClient(fl.client.NumPyClient):
+    """Flower client for federated learning.
+
+    Args:
+        cid (str): The client ID.
+        trainloader (torch.utils.data.DataLoader): DataLoader for training data.
+        testloader (torch.utils.data.DataLoader): DataLoader for test data.
+        valloader (torch.utils.data.DataLoader): DataLoader for validation data
+        cfg (Config): Hydra Configuration.
+        local_layers (list): List of local layers.
+        local_optims (list): List of local optimizers.
+        local (bool): Whether to use local layers.
+
+    Attributes
+    ----------
+        cid (str): The client ID.
+        trainloader (torch.utils.data.DataLoader): DataLoader for training data.
+        testloader (torch.utils.data.DataLoader): DataLoader for test data.
+        valloader (torch.utils.data.DataLoader): DataLoader for validation data
+        local_layers (list): List of local layers.
+        local_optims (list): List of local optimizers.
+        local (bool): Whether to use local layers.
+        device (torch.device): The device to run the model on.
+        epochs (int): Number of training epochs.
+        n_kernels (int): Number of convolutional kernels.
+        learning_rate (float): Learning rate.
+        weight_decay (float): Weight decay.
+        net (CNNTarget): Target neural network model.
+    """
+
+    # pylint: disable=too-many-arguments
+    def __init__(
+        self,
+        cid,
+        trainloader,
+        testloader,
+        valloader,
+        cfg,
+        local_layers,
+        local_optims,
+        local,
+    ) -> None:
+        super().__init__()
+
+        self.cid = cid
+        self.trainloader = trainloader
+        self.testloader = testloader
+        self.valloader = valloader
+        self.local_layers = local_layers
+        self.local_optims = local_optims
+        self.local = local
+        self.gpus = cfg.client_resources.gpus
+        self.device = get_device(gpus=self.gpus)
```

Comment:
```suggestion
        self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```

Reply:
the device should be chosen automatically based on what resources the `VirtualClientEngine` assigns to a client. So the above will use `cuda` if `client_resources.num_gpus > 0`

Reply:
Taken care